sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/StorageSeekableByteChannelShareFileReadBehavior.java

```diff
@@ -0,0 +1,84 @@
+import com.azure.storage.file.share.models.ShareRequestConditions;
+import com.azure.storage.file.share.models.ShareStorageException;
+import com.azure.storage.file.share.options.ShareFileDownloadOptions;
+import com.fasterxml.jackson.databind.util.ByteBufferBackedOutputStream;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+class StorageSeekableByteChannelShareFileReadBehavior implements StorageSeekableByteChannel.ReadBehavior {
+    private static final ClientLogger LOGGER = new ClientLogger(StorageSeekableByteChannelShareFileReadBehavior.class);
+
+    private final ShareFileClient client;
+    private final ShareRequestConditions conditions;
+
+    private Long lastKnownResourceLength;
+
+    StorageSeekableByteChannelShareFileReadBehavior(ShareFileClient client, ShareRequestConditions conditions) {
+        this.client = client;
+        this.conditions = conditions;
+    }
+
+    ShareFileClient getClient() {
+        return this.client;
+    }
+
+    ShareRequestConditions getRequestConditions() {
+        return this.conditions;
+    }
+
+    @Override
+    public int read(ByteBuffer dst, long sourceOffset) {
+        if (dst.remaining() <= 0) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException("'dst.remaining()' must be positive."));
+        }
+
+        int initialPosition = dst.position();
+
+        try (ByteBufferBackedOutputStream dstStream = new ByteBufferBackedOutputStream(dst)) {
+            ShareFileDownloadResponse response =  client.downloadWithResponse(dstStream,
+                new ShareFileDownloadOptions()
+                    .setRange(new ShareFileRange(sourceOffset, sourceOffset + dst.remaining() - 1))
+                    .setRequestConditions(conditions),
+                null, null);
+            lastKnownResourceLength = getResourceLengthFromContentRange(
+                response.getDeserializedHeaders().getContentRange());
+            return dst.position() - initialPosition;
+        } catch (ShareStorageException e) {
+            if (e.getErrorCode() == ShareErrorCode.INVALID_RANGE) {
+                String contentRange = e.getResponse().getHeaderValue("Content-Range");
+                if (contentRange != null) {
+                    lastKnownResourceLength = getResourceLengthFromContentRange(contentRange);
+                }
+                // if requested offset is past updated end of file, then signal end of file. Otherwise, only signal
+                // that zero bytes were read
+                return sourceOffset < lastKnownResourceLength ? 0 : -1;
+            }
+            throw LOGGER.logExceptionAsError(e);
+        } catch (IOException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    @Override
+    public long getCachedLength() {
+        if (lastKnownResourceLength == null) {
+            lastKnownResourceLength = client.getProperties().getContentLength();
```

Comment:
Admittedly I am not familiar with java conventions or the `SeekableByteChannel` concept but to me it does feel a little odd that something that is getting a _cached_ length would make a service call?

Is this used somewhere in our functionality or just exposed as part of the interface? If it's not need by us, maybe we consider just throwing if you calling and we have no length? Again, not sure what the convention would typically be if nothing is cached. Feel free to ignore, just calling out that it seems a little weird to me.

Reply:
No this is valid. Originally, it was not going to make a service call, but that led to the situation where a call to `SeekableByteChannel.size()` could throw if it was made upfront, which is not desirable behavior. This API can probably be renamed to `getResourceLength()` and the expectation the channel makes is that the behavior could return a cached value instead of making a service call.

The functionality isn't present in the files service, but for something like blobs, the use of etag locking by default would lead to the cached value never being out of date, as if it were, any read would throw.

Reply:
Ok, a rename could help for sure. I was mainly getting hung up on the "cached" part fetching a value, but I think it's reasonable for something like `getResourceLength` to use a cached value if present. Didn't know if this was an overload for a standard lib function or something we control though.